src/anim/controller/anim-blend-tree.js

```diff
@@ -84,112 +66,20 @@ class AnimBlendTree extends AnimNode {
-                }
-                for (i = 0; i < this._children.length; i++) {
-                    this._children[i].weight = this._children[i]._weight / weightSum;
-                }
-                break;
-            }
-            case ANIM_BLEND_2D_DIRECTIONAL: {
-                parameterValues = this._parameters.map(function (param) {
-                    return this._findParameter(param).value;
-                }.bind(this));
-                if (this._parameterValues && (JSON.stringify(this._parameterValues) === JSON.stringify(parameterValues))) {
-                    return;
-                }
-                this._parameterValues = parameterValues;
-                p = new Vec2(this._parameterValues);
-
-
-                weightSum = 0.0;
-                for (i = 0; i < this._children.length; i++) {
-                    pi = this._children[i].point.clone();
-                    minj = Number.MAX_VALUE;
-                    for (j = 0; j < this._children.length; j++) {
-                        if (i === j) continue;
-                        pj = this._children[j].point.clone();
-                        var pipAngle = getAngleRad(pi, p);
-                        var pipjAngle = getAngleRad(pi, pj);
-                        pipj = new Vec2((pj.length() - pi.length()) / ((pj.length() + pi.length()) / 2), pipjAngle * 2.0);
-                        pip = new Vec2((p.length() - pi.length()) / ((pj.length() + pi.length()) / 2), pipAngle * 2.0);
-                        result = clamp(1.0 - Math.abs((pip.clone().dot(pipj) / pipj.lengthSq())), 0.0, 1.0);
-                        if (result < minj) minj = result;
-                    }
-                    this._children[i].weight = minj;
-                    weightSum += minj;
-                }
-                for (i = 0; i < this._children.length; i++) {
-                    this._children[i].weight = this._children[i]._weight / weightSum;
-                }
-                break;
-            }
-            case ANIM_BLEND_DIRECT: {
-                parameterValues = this._parameters.map(function (param) {
-                    return this._findParameter(param).value;
-                }.bind(this));
-                if (this._parameterValues === parameterValues) {
-                    return;
-                }
-                this._parameterValues = parameterValues;
-                weightSum = 0.0;
-                for (i = 0; i < this._children.length; i++) {
-                    weightSum += clamp(this._parameterValues[i], 0.0, Number.MAX_VALUE);
-                }
-                for (i = 0; i < this._children.length; i++) {
-                    this._children[i].weight = clamp(this._parameterValues[i], 0.0, Number.MAX_VALUE) / weightSum;
-                }
-                break;
+    currentParameterValues() {
+        return this._parameters.map(function (param) {
+            return this._findParameter(param).value;
+        }.bind(this));
+    }
```

Comment:
each time this gets called, and array is allocated .. I think class should store single array and this function should just update it.

Reply:
Done 👍